### PR TITLE
feat(pub): Support updating lock files

### LINKED ISSUE #6049: Renovate not updating pubspec.lock
**What Renovate type are you using?**

Self-hosted, with `--binary-source docker`.

**Describe the bug**

Renovate is opening pull requests that update `pubspec.yaml`, but which do not update `pubspec.lock` with corresponding changes.

**Did you see anything helpful in debug logs?**

Nothing specific, but here is the generation of one PR, to upgrade the dependency `flutter_typeahead`:

<details><summary>Logs</summary>

```
DEBUG: processBranch with 1 upgrades (repository=tidbyt/tidbyt, dependencies=flutter_typeahead, branch=renovate/flutter_typeahead-1.x)
--
  | DEBUG: Setting baseBranch to master (repository=tidbyt/tidbyt, branch=renovate/flutter_typeahead-1.x)
  | DEBUG: latest commit (repository=tidbyt/tidbyt, branch=renovate/flutter_typeahead-1.x)
  | "branchName": "master",
  | "latestCommitDate": "2020-04-24 16:30:42 -0400"
  | DEBUG: getBranchPr(renovate/flutter_typeahead-1.x) (repository=tidbyt/tidbyt, branch=renovate/flutter_typeahead-1.x)
  | DEBUG: findPr(renovate/flutter_typeahead-1.x, undefined, open) (repository=tidbyt/tidbyt, branch=renovate/flutter_typeahead-1.x)
  | DEBUG: branchExists=false (repository=tidbyt/tidbyt, branch=renovate/flutter_typeahead-1.x)
  | DEBUG: Branch has been checked in master issue: unlimit (repository=tidbyt/tidbyt, branch=renovate/flutter_typeahead-1.x)
  | DEBUG: Branch has 1 upgrade(s) (repository=tidbyt/tidbyt, branch=renovate/flutter_typeahead-1.x)
  | DEBUG: recreateClosed is false (repository=tidbyt/tidbyt, branch=renovate/flutter_typeahead-1.x)
  | DEBUG: findPr(renovate/flutter_typeahead-1.x, fix(deps): update dependency flutter_typeahead to v1.8.0, !open) (repository=tidbyt/tidbyt, branch=renovate/flutter_typeahead-1.x)
  | DEBUG: prAlreadyExisted=false (repository=tidbyt/tidbyt, branch=renovate/flutter_typeahead-1.x)
  | DEBUG: Checking schedule(at any time, null) (repository=tidbyt/tidbyt, branch=renovate/flutter_typeahead-1.x)
  | DEBUG: No schedule defined (repository=tidbyt/tidbyt, branch=renovate/flutter_typeahead-1.x)
  | DEBUG: Branch needs creating (repository=tidbyt/tidbyt, branch=renovate/flutter_typeahead-1.x)
  | DEBUG: Using parentBranch: undefined (repository=tidbyt/tidbyt, branch=renovate/flutter_typeahead-1.x)
  | DEBUG: manager.getUpdatedPackageFiles() (repository=tidbyt/tidbyt, branch=renovate/flutter_typeahead-1.x)
  | DEBUG: Starting search at index 861 (repository=tidbyt/tidbyt, branch=renovate/flutter_typeahead-1.x)
  | DEBUG: Found match at index 861 (repository=tidbyt/tidbyt, branch=renovate/flutter_typeahead-1.x)
  | DEBUG: Contents updated (repository=tidbyt/tidbyt, branch=renovate/flutter_typeahead-1.x)
  | DEBUG: Updated 1 package files (repository=tidbyt/tidbyt, branch=renovate/flutter_typeahead-1.x)
  | DEBUG: Getting updated lock files (repository=tidbyt/tidbyt, branch=renovate/flutter_typeahead-1.x)
  | DEBUG: lock file dirs (repository=tidbyt/tidbyt, branch=renovate/flutter_typeahead-1.x)
  | "dirs": {
  | "yarnLockDirs": [],
  | "npmLockDirs": [],
  | "pnpmShrinkwrapDirs": [],
  | "lernaDirs": []
  | }
  | DEBUG: Writing package.json files (repository=tidbyt/tidbyt, branch=renovate/flutter_typeahead-1.x)
  | "packageFiles": ["package.json"]
  | DEBUG: Writing package-lock.json (repository=tidbyt/tidbyt, branch=renovate/flutter_typeahead-1.x)
  | DEBUG: Writing any updated package files (repository=tidbyt/tidbyt, branch=renovate/flutter_typeahead-1.x)
  | DEBUG: No updated lock files in branch (repository=tidbyt/tidbyt, branch=renovate/flutter_typeahead-1.x)
  | DEBUG: Checking for post-upgrade tasks (repository=tidbyt/tidbyt, branch=renovate/flutter_typeahead-1.x)
  | "tasks": {"commands": [], "fileFilters": []},
  | "allowedCommands": [".*"]
  | DEBUG: 1 file(s) to commit (repository=tidbyt/tidbyt, branch=renovate/flutter_typeahead-1.x)
  | DEBUG: Committing files to branch renovate/flutter_typeahead-1.x (repository=tidbyt/tidbyt, branch=renovate/flutter_typeahead-1.x)
```

</details>

**To Reproduce**

This is a private monorepo but I'm happy to provide access temporarily if it helps with debugging.

**Additional context**

This is a monorepo with the Dart project living in the `flutter/` directory, so the relevant files are `flutter/pubspec.yaml` and `flutter/pubspec.lock`. Perhaps this is similar to #2519?

**blocked by**
- [x] #12820 
- [x] https://github.com/containerbase/buildpack/issues/200

### PR DESCRIPTION
**Pull request by zeshuaro**
## Changes

Add support to also update `pubspec.lock` when a Pub dependency is updated

## Context

- Closes #6049 
- Closes #18412
- Continuation of #18412 as the author [mentioned](https://github.com/renovatebot/renovate/pull/18412#issuecomment-1327083860) that they don't time to finish the work
- The original commit on the PR is retained, I only added one additional commit to address the comments outlined [here](https://github.com/renovatebot/renovate/pull/18412#discussion_r1016643204)

## Documentation

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository
  - Sample PR with the lock file being updated - https://github.com/barriot/r0/pull/6

### CODE REVIEW

File: lib/modules/manager/pub/artifacts.spec.ts [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+            HOME: '/home/user',
+            HTTPS_PROXY: 'https://example.com',
+            HTTP_PROXY: 'http://example.com',
+            LANG: 'en_US.UTF-8',
+            LC_ALL: 'en_US',
+            NO_PROXY: 'localhost',
+            PATH: '/tmp/path',
+          },
+          maxBuffer: 10485760,
+          timeout: 900000,
```
Comment (viceice) [inline lib/modules/manager/pub/artifacts.spec.ts:86]:
```suggestion
```
remove those values, they are same for all

same for other tests

**Review by viceice — COMMENTED** [review]
needs tests for install and docker mode. checkout cargo or nugget manager for samples

File: lib/modules/manager/pub/artifacts.spec.ts (hunk 1 above)
Comment (zeshuaro) [reply lib/modules/manager/pub/artifacts.spec.ts:86]: ### Docker mode tests

Hopefully I'm on the right direction for the docker mode tests. I tracked down that we probably need to update `allToolConfig` in [containerbase](https://github.com/renovatebot/renovate/blob/main/lib/util/exec/containerbase.ts#L17) and define the tools `Dart` and `Flutter`.

For Flutter, I think the config would be something like this, which seems to be working:

```typescript
flutter: {
  datasource: 'flutter-version',
  depName: 'flutter',
  versioning: npmVersioningId,
},
```

But for Dart, we don't have a data source for it, so I think we would need to add a new data source for Dart. Potentially something similar to this issue: https://github.com/renovatebot/renovate/issues/12820. Would you be able to confirm? If this is the case, I'm also happy to work on adding a data source for Dart.

### Install mode tests

For the install mode tests, I've updated the `GlobalConfig` but it's still not running in install mode (the `exec` snapshot doesn't contain the `install-tool` command). Is there some other config that I need to specify?

<details>
  <summary>Test</summary>
  
  ```ts
  it('supports install mode', async () => {
    GlobalConfig.set({ ...adminConfig, binarySource: 'install' });
    const execSnapshots = mockExecAll();
    fs.getSiblingFileName.mockReturnValueOnce('pubspec.lock');
    fs.readLocalFile.mockResolvedValueOnce('Old pubspec.lock');
    fs.readLocalFile.mockResolvedValueOnce('New pubspec.lock');
    expect(
      await pub.updateArtifacts({
        packageFileName: 'pubspec.yaml',
        updatedDeps: [{ depName: 'dep1' }],
        newPackageFileContent: 'sdk: flutter',
        config,
      })
    ).toEqual([
      {
        file: {
          type: 'addition',
          path: 'pubspec.lock',
          contents: 'New pubspec.lock',
        },
      },
    ]);
    expect(execSnapshots).toMatchObject([
      { cmd: 'install-tool flutter 3.3.9' },
      { cmd: 'flutter pub upgrade' },
    ]);
  });
  ```
</details>

**Comment by zeshuaro** [comment]
### Docker mode tests

Hopefully I'm on the right direction for the docker mode tests. I tracked down that we probably need to update `allToolConfig` in [containerbase](https://github.com/renovatebot/renovate/blob/main/lib/util/exec/containerbase.ts#L17) and define the tools `Dart` and `Flutter`.

For Flutter, I think the config would be something like this, which seems to be working:

```typescript
flutter: {
  datasource: 'flutter-version',
  depName: 'flutter',
  versioning: npmVersioningId,
},
```

But for Dart, we don't have a data source for it, so I think we would need to add a new data source for Dart. Potentially something similar to this issue: https://github.com/renovatebot/renovate/issues/12820. Would you be able to confirm? If this is the case, I'm also happy to work on adding a data source for Dart.

### Install mode tests

For the install mode tests, I've updated the `GlobalConfig` but it's still not running in install mode (the `exec` snapshot doesn't contain the `install-tool` command). Is there some other config that I need to specify?

<details>
  <summary>Test</summary>
  
  ```ts
  it('supports install mode', async () => {
    GlobalConfig.set({ ...adminConfig, binarySource: 'install' });
    const execSnapshots = mockExecAll();
    fs.getSiblingFileName.mockReturnValueOnce('pubspec.lock');
    fs.readLocalFile.mockResolvedValueOnce('Old pubspec.lock');
    fs.readLocalFile.mockResolvedValueOnce('New pubspec.lock');
    expect(
      await pub.updateArtifacts({
        packageFileName: 'pubspec.yaml',
        updatedDeps: [{ depName: 'dep1' }],
        newPackageFileContent: 'sdk: flutter',
        config,
      })
    ).toEqual([
      {
        file: {
          type: 'addition',
          path: 'pubspec.lock',
          contents: 'New pubspec.lock',
        },
      },
    ]);
    expect(execSnapshots).toMatchObject([
      { cmd: 'install-tool flutter 3.3.9' },
      { cmd: 'flutter pub upgrade' },
    ]);
  });
  ```
</details>

**Comment by rarkins** [comment]
Yes on `allToolConfig`. Consider `semver` versioning instead of `npm` if you don't need ranges, or keep as `npm` if you need npm's range syntax.

For Dart, you could maybe use https://github.com/dart-lang/sdk/tags although you want to make sure that it matches https://github.com/containerbase/base/blob/7c5021bf3f9c39e6513fe932becd52ee349057d8/src/usr/local/buildpack/tools/v2/dart.sh#L16

**Comment by zeshuaro** [comment]
> For Dart, you could maybe use https://github.com/dart-lang/sdk/tags although you want to make sure that it matches https://github.com/containerbase/base/blob/7c5021bf3f9c39e6513fe932becd52ee349057d8/src/usr/local/buildpack/tools/v2/dart.sh#L16

I remember Google often pushes the new tag first but then uploads the SDK to their storage site a few days later (this happened for Flutter at least). So it seems like a new data source for Dart would be preferable?

**Comment by rarkins** [comment]
Yes, it would be preferable

File: lib/modules/manager/pub/artifacts.spec.ts [hunk 2]
```diff
@@ -0,0 +1,265 @@
+import { join } from 'upath';
```
Comment (HonkingGoose) [inline lib/modules/manager/pub/artifacts.spec.ts:1]: Please put the correct lockfile name and format for the `pub` manager in the list here:

https://github.com/renovatebot/renovate/blob/66159ac549e4ecb8e2425193b1d40bb368576bb0/docs/usage/configuration-options.md?plain=1#L1413-L1424

**Comment by zeshuaro** [comment]
Sorry if you received a couple of review re-requesting, GitHub doesn't let me to re-request both of you for some reasons.

**Comment by HonkingGoose** [comment]
> Sorry if you received a couple of review re-requesting, GitHub doesn't let me to re-request both of you for some reasons.

There's a "refresh" icon you can use to re-request a review, see step 7 in the GitHub docs. Did that not work properly for you?

https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review

**Review by HonkingGoose — DISMISSED** [review]
I'm happy with the docs.

**Comment by zeshuaro** [comment]
It did work, but it wouldn't let me to re-request reviews from both of you. So right now I requested one of you. If I request for the other person, GitHub would remove the review request from the first person 😓

File: lib/modules/manager/pub/artifacts.ts [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    const toolName = isFlutter ? 'flutter' : 'dart';
+
+    const constraint = isFlutter
+      ? config.constraints?.flutter ?? getFlutterConstraint(oldLockFileContent)
+      : config.constraints?.dart ?? getDartConstraint(oldLockFileContent);
+    await exec(`${toolName} pub upgrade`, {
+      cwdFile: packageFileName,
+      docker: {
+        image: 'sidecar',
+      },
```
Comment (viceice) [inline lib/modules/manager/pub/artifacts.ts:58]:
```suggestion
      docker: {},
```
image is now managed centrally

File: lib/modules/manager/pub/artifacts.spec.ts [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+          cmd:
+            'docker run --rm --name=renovate_sidecar --label=renovate_child ' +
+            '-v "/tmp/github/some/repo":"/tmp/github/some/repo" ' +
+            '-v "/tmp/cache":"/tmp/cache" ' +
+            '-e BUILDPACK_CACHE_DIR ' +
+            '-e CONTAINERBASE_CACHE_DIR ' +
+            '-w "/tmp/github/some/repo" ' +
+            'renovate/sidecar ' +
+            'bash -l -c ' +
+            `"install-tool ${params.sdk} 3.3.9 && ${params.sdk} pub upgrade"`,
```
Comment (viceice) [inline lib/modules/manager/pub/artifacts.spec.ts:179]: move the quotes to the previous part and also split commands. checkout other managers for samples

**Review by viceice — APPROVED** [review]

**Comment by provokateurin** [comment]
Awesome work :heart: Thanks a lot for taking over @rarkins !

**Comment by rarkins** [comment]
No, thanks to @zeshuaro !

**Comment by provokateurin** [comment]
Ooops I didn't check the Github UI correctly

**Pull request merged**